src/ham/hamctl/subsys-accounts.cpp

```diff
@@ -1,76 +1,213 @@
+        return;
+    }
+
+    if (esc.state == escaped)
+    {
+        esc.state = c == '[' ? square : normal;
+        return;
+    }
+
+    if (esc.state == square)
+    {
+        esc.state = getpars;
+        esc.npar = 0;
+        if (c == '?')
+            return;
+    }
+
+    if (esc.state == getpars)
+    {
+        if (c == ';' && esc.npar < (ESCAPE_NPAR - 1))
+        {
+            esc.npar++;
+            return;
+        }
+        if (c >= '0' && c <= '9')
+            return;
+    }
 
-    (void) tcsetattr(STDIN_FILENO, TCSANOW, &tty);
+    esc.state = normal;
 }
 
-static void read_pw(char * pw)
+/**
+ * @brief Read password from stdin. The password gets obscured with '*'
+ *        characters as it is being typed.
+ *
+ * @param[OUT] pw Where password will be saved.
+ */
+void read_pw(char *pw)
 {
-    char      c;
-    unsigned  i = 0;
-    while (std::cin.get(c))
+    char        c;
+    unsigned    i = 0;
+    esc_data_c  esc_data;
+
+    // Disable STDIN echo to obscure password while it is being typed
+    struct termios oflags, nflags;
+
+    tcgetattr(STDIN_FILENO, &oflags);
+    nflags = oflags;
+
+    nflags.c_lflag &= ~(ICANON | ECHO);
+    nflags.c_cc[VTIME] = 0;
+    nflags.c_cc[VMIN] = 1;
+
+    (void)tcsetattr(STDIN_FILENO, TCSANOW/*TCSAFLUSH*/, &nflags);
+
+    while (std::cin.get(c) && (c != '\n'))
     {
-        if (c == '\n')
+        if (esc_data.state != normal)
         {
-            pw[i] = '\0';
+            escape_sequence(esc_data, c);
+            continue;
+        }
+
+        switch (c)
+        {
+        case '\b':
+        case DEL:  // backspace
+            if (i > 0)
+            {
+                pw[--i] = '\0';
+                std::cout << "\b \b";
+            }
+            continue;
+
+        case ESC:
+            escape_sequence(esc_data, c);
             break;
+
+        default:
+            if (isprint(c) && !isspace(c))
+            {
+                pw[i++] = c;
+                std::cout << '*';
+            }
         }
-        pw[i++] = c;
-        std::cout << '*' << std::flush; // TODO: This is not working. It looks like buffering is enabled on the stdout
     }
-    std::cout.put('\n');
+    std::cout << std::endl;
+
+    // Restore STDIN config (i.e. echo)
+    (void)tcsetattr(STDIN_FILENO, TCSANOW, &oflags);
+
+    pw[i] = '\0';
 }
 
+/**
+ * @brief Generate a random SHA-512 salt to be used when invoking crypt_r()
+ *
+ * @return The salt string in the form "$6$random-salt-string$"
+ */
+static std::string get_salt()
+{
+    std::string  salt;
+    unsigned     salt_len = 0;         // Salt can be at most 16 chars long
+    while (salt_len < 4)               // Make sure salt is at least 4 chars long
+        salt_len = 1 + (rand() & 0xF); // Yields a value in the range 1..16
+
+    while (salt.length() < salt_len)
+    {
+        char c = (char)rand();
+        if (isprint(c) && (c != '\\') && (c != '$'))
```

Comment:
My understanding of the salt is that it is supposed to be base64 encoded. My suggestion is to replace this condition with `isalnum(c) && isascii(c)` (isascii to ensure that locale settings don't add additional characters outside `[A-Z a-z 0-9]`)

Reply:
In  `**cryp_t**()`'s documentation, they only say that `**salt**` should be a string of at most 16 chars. The `**$**` char is reserved as a delimiter. I also avoided the **backslash** because that gets interpreted as an escape character. All other characters seem to be fine. That is, `**crypt_r**()` does not complain with characters outside of `[A-Z a-z 0-9]`. The `**salt**` is used simply for additional scrambling before hashing the data . And allowing more chars than just `[A-Z a-z 0-9]` will result in even more scrambling. 

I will search the Internet to see what they say about `**crypt_r**()` and what is recommended for `**salt**`. I will get back to you on this.

Reply:
I found the source code for `**mkpasswd**` and in that code they do limit the char set to "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789./"

So I shall do the same and submit a new pull request